classes/boardmanager.php

```diff
@@ -0,0 +1,1152 @@
+
+            $newboard['id'] = $DB->insert_record('kanban_board', $newboard);
+            $columns = $DB->get_records('kanban_column', ['kanban_board' => $template->id]);
+            $cards = $DB->get_records('kanban_card', ['kanban_board' => $template->id]);
+            $newcolumn = [];
+            $newcard = [];
+            foreach ($columns as $column) {
+                $newcolumn[$column->id] = clone $column;
+                $newcolumn[$column->id]->kanban_board = $newboard['id'];
+                $newcolumn[$column->id]->timecreated = time();
+                $newcolumn[$column->id]->timemodified = time();
+                unset($newcolumn[$column->id]->id);
+                $newcolumn[$column->id]->id = $DB->insert_record('kanban_column', $newcolumn[$column->id]);
+            }
+            foreach ($cards as $card) {
+                $newcard[$card->id] = clone $card;
+                $newcard[$card->id]->kanban_board = $newboard['id'];
+                $newcard[$card->id]->timecreated = time();
+                $newcard[$card->id]->timemodified = time();
+                $newcard[$card->id]->kanban_column = $newcolumn[$card->kanban_column]->id;
+                $newcard[$card->id]->originalid = $card->id;
+                unset($newcard[$card->id]->id);
+                // Remove user id of original creator.
+                unset($newcard[$card->id]->createdby);
+                $newcard[$card->id]->id = $DB->insert_record('kanban_card', $newcard[$card->id]);
+                // Copy attachment files.
+                if ($context) {
+                    $this->copy_attachment_files($context->id, $card->id, $newcard[$card->id]->id);
+                }
+            }
+
+            $newboard['sequence'] = helper::sequence_replace($newboard['sequence'], $newcolumn);
+            $DB->update_record('kanban_board', $newboard);
+            helper::update_cached_board($newboard['id']);
+            foreach ($newcolumn as $col) {
+                $col->sequence = helper::sequence_replace($col->sequence, $newcard);
+                $DB->update_record('kanban_column', $col);
+            }
+            return $newboard['id'];
+        }
+    }
+
+    /**
+     * Deletes a board and all contents of it.
+     *
+     * @param int $id The board id
+     * @return void
+     */
+    public function delete_board(int $id) {
+        global $DB;
+        // Cards need to be read to identify files, assignees and discussions.
+        $cardids = $DB->get_fieldset_select('kanban_card', 'id', 'kanban_board = :id', ['id' => $id]);
+        $this->delete_cards($cardids);
+
+        $DB->delete_records('kanban_history', ['kanban_board' => $id]);
+        $DB->delete_records('kanban_column', ['kanban_board' => $id]);
+        $DB->delete_records('kanban_card', ['kanban_board' => $id]);
+        $DB->delete_records('kanban_board', ['id' => $id]);
+        // The rest of the elements is skipped in the update message.
+        $this->load_board($id);
```

Comment:
Does this even work after everything is being deleted?

Should the cache be cleared/reset here?

Reply:
Yes - this loads only board record which is still existing. There is no need to clear the cache as the board id is not reused.

Reply:
alright. Clearing cache would be nice anyway, just to cleanup what has been stored in the cache, right? As there is no automatic invalidation time the board would continue to live forever in the cache (until the cache is being cleared/restarted/whatever of course). But still would be nice to cleanup probably?